### ISSUE TITLE
Training, validation an test each in separate sessions and own graphs

### ISSUE THREAD

**Issue by tilmankamp** [mozilla/DeepSpeech#176]

**Comment by Cwiiis** [comment]
My branch for #108 ends up doing this, though it's quite messy at the moment - test is already in a separate session, then it splits train/validation into separate sessions using a placeholder + cond to pick the data set (which allows them to share a graph). I wonder what's more efficient out of separate graphs (and thus separate sessions) vs. shared graphs + cond and separate sessions... (or one huge graph as we have right now)

**Comment by tilmankamp** [comment]
There are basically two competing solutions to not having multiple graphs at the same time on the GPU:
- Having a single graph plus conditionals
- Producing different graphs one after the other by building and closing sessions
This issue is about the latter one and meant as an experiment to figure out, if this will lead to higher GPU utilization (as we observed conditional placing on the CPU).

We should actually measure both approaches.

**Comment by Cwiiis** [comment]
The duplication is fixed in #192;
Master: https://www.dropbox.com/s/mlir5523m84mw0d/graph-master.png?dl=0
#192: https://www.dropbox.com/s/h1v07npdepy4yyg/graph-no-duplication.png?dl=0

I don't know if it affects performance yet - a master run had duration 35 and my branch had 32, but that's not a significant amount for such a small run (ldc). Need to run this on a couple of iterations of ted on the training machine.

**Comment by tilmankamp** [comment]
This is a great leap ahead. Once you PR gets closed, I'll close this one as irrelevant.

**Comment by tilmankamp** [comment]
Got invalid as #192 got merged.

**Comment by tilmankamp** [comment]
Following tensorflow developer ebrevdo's advice in [[1]](https://groups.google.com/a/tensorflow.org/d/msg/discuss/mLrt5qc9_uU/sGNbC7GpAwAJ), we should give this a spin.

The refactored main loop will probably merge display-, checkpoint- and validation-steps, as the only way to continue an interrupted (training) session is by checkpoints.

@reuben @kdavis-mozilla @Cwiiis @lissyx What do you think?

**Comment by kdavis-mozilla** [comment]
The idea seems reasonable to me.

However, just to make sure we're on the same page could you give a little psudocode (Nothing too detailed!) indicating how the new refactored code will look.

**Comment by tilmankamp** [comment]
```
train_session = tf.Session(...)
for epoch in xrange(train_iters):
    for batch in xrange(...):
        results = train_session.run(...); # training 
        collect_results(results_array, results)
    if checkpoint_step:
        calculate_and_print_wer_report(results_array)
        path = save_session(train_session) #  will also be used as general checkpoint
        train_session.close()
        dev_session = tf.Session(...) # could also be part of run_inference
        dev_wer = run_inference(dev_session, "Validation", ...)
        dev_session.close() # could also be part of run_inference
        train_session = load_session(path)
train_session.close()
test_session = tf.Session(...) # could also be part of run_inference
test_wer = run_inference(test_session, "Test", ...)
test_session.close() # could also be part of run_inference
```

**Comment by kdavis-mozilla** [comment]
Seems reasonable to me.

The only thing which is slightly worrying is that now we have to tie checkpointing and WER reports together.

In general checkpointing is relatively lightweight but WER reports are not. So, we'll have to make fewer checkpoints or run slower.

**Comment by Cwiiis** [comment]
Would it make some sense to run through all the training steps first, with checkpointing of course, then do the dev runs afterwards? (and still allow wer reporting not to be done on every step of checkpointing/validation)

**Comment by tilmankamp** [comment]
@kdavis-mozilla Fair point - this is how it could look like:
```Python
train_session = tf.Session(...)
for epoch in xrange(train_iters):
    for batch in xrange(...):
        results = train_session.run(...); # training 
        collect_results(results_array, results)
    if display_step:
        calculate_and_print_wer_report(results_array)
    if checkpoint_step:
        path = save_session(train_session) #  will also be used as general checkpoint
        train_session.close()
        dev_session = tf.Session(...) # could also be part of run_inference
        dev_wer = run_inference(dev_session, "Validation", ...)
        dev_session.close() # could also be part of run_inference
        train_session = load_session(path)
train_session.close()
test_session = tf.Session(...) # could also be part of run_inference
test_wer = run_inference(test_session, "Test", ...)
test_session.close() # could also be part of run_inference
```

**Comment by tilmankamp** [comment]
@Cwiiis  So you want to run them on checkpoint-ed versions of the model?

**Comment by kdavis-mozilla** [comment]
The validation runs are used to monitor training while training is ongoing.

For example if at some point in the training run the loss of the training data set is 0, lower loss is better, but the validation loss is 500, then one knows that the training run is overfitting the training data and the run should be stopped immediately and `dropout_rate`, the means of regularization we are using,  should be increased.

**Comment by Cwiiis** [comment]
I was thinking though, that we'd likely get through many epochs before validation becomes useful. For efficiency, I think we might find that we get to where we want quicker if we do all the training first, then run validation separately, backwards from the last checkpoint. These might be bad assumptions, though, and they could subject us to local maxima/minima if we acted too quickly on the results I suppose.

**Comment by tilmankamp** [comment]
As an alternative we could introduce some non-linear progression on validation step length.

**Comment by kdavis-mozilla** [comment]
@Cwiiis what if we train 350 epochs, not an unreasonable number[[1](http://bit.ly/2gur8YO)], but then look backward and find that we could have stopped at epoch 30 due to overfitting if we had looked at the validation results?

We will have wasted 320 epochs which right now is about 20 days of training.

It seems like what you want to add is a parameter `min_validation_epoch` which could be used something like
```python
is_validation_step = validation_step > 0 and \
                                 (epoch + 1) % validation_step == 0 and \
                                 epoch > min_validation_epoch
```
We can figure out a reasonable value for `min_validation_epoch` by looking at our current training run(s).

**Comment by kdavis-mozilla** [comment]
@tilmankamp @Cwiiis Tilman I guess we basically have the same idea.

**Comment by Cwiiis** [comment]
That sounds good :) Maybe that could be based on loss rather than a set number of runs (or both in conjunction somehow)? We probably only need to worry about overfitting once loss reaches a particular level.

**Comment by kdavis-mozilla** [comment]
@Cwiiis Yeah loss sounds like a good idea as it's something we calculate all the time anyway, thus no computational overhead.

**Comment by tilmankamp** [comment]
@Cwiiis @kdavis-mozilla Just #224 for this.

**Comment by tilmankamp** [comment]
@kdavis-mozilla @Cwiiis But at first I'll work on the checkpoints based training loop.
